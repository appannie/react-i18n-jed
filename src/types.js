// @flow
export type I18nType = {
    lang: string,
    gettext: string => string,
    pgettext: (string, string) => string,
    ngettext: (string, string, number) => string,
    npgettext: (string, string, string, number) => string,
};
