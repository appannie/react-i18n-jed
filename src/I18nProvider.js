// @flow
import React from 'react';
import PropTypes from 'prop-types';

import { type I18nType } from '.';

class I18nProvider extends React.Component {
    props: {
        i18n: I18nType,
        children?: any,
    };

    getChildContext() {
        const { i18n } = this.props;
        return { i18n };
    }

    render() {
        const { children } = this.props;
        return React.Children.only(children);
    }
}

I18nProvider.childContextTypes = {
    i18n: PropTypes.object,
};

export default I18nProvider;
