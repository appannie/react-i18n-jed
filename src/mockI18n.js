// @flow
const mockSingular = (key: string) => key;
const mockContextSingular = (context: string, key: string) => key;
const mockPlural = (key: string, pluralKey: string, count: number) => {
    if (count === 1) {
        return key;
    }
    return pluralKey;
};

const mockI18n = {
    lang: 'en-US',
    gettext: mockSingular,
    ngettext: mockPlural,
    pgettext: mockContextSingular,
    npgettext: (context: string, key: string, pluralKey: string, count: number) =>
        mockPlural(key, pluralKey, count),
};

export default mockI18n;
