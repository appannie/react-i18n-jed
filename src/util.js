// @flow
import Jed from 'jed';
import fetchPonyfill from 'fetch-ponyfill';
import Cookies from 'universal-cookie';

const langJSONFileName = 'aa_react.json';
const { fetch } = fetchPonyfill();
export const LANG_MAP = {
    en: 'en-US',
    de: 'de-DE',
    fr: 'fr-FR',
    jp: 'ja-JP',
    kr: 'ko-KR',
    ru: 'ru-RU',
    cn: 'zh-CN',
};

const checkStatus = (res: Response) => {
    if (res.ok) {
        return res;
    }

    throw new Error(`Fetch error: ${res.url} returned ${res.status} ${res.statusText}`);
};
const parseJSON = (res: Response) => res.json();

function getI18n(localeJSON: Object) {
    const i18n = new Jed(localeJSON);
    return i18n;
}

async function getLocale(req?: Object) {
    let baseUrl = '';
    const cookies = new Cookies(req ? req.headers.cookie || '' : null);
    const aaLangName = cookies.get('aa_language');
    const lang = LANG_MAP[aaLangName] || 'en-US';

    if (req && req.connection) {
        const address = req.socket.server.address();
        const nodeUrlProtocol = req.connection.encrypted ? 'https:' : 'http:';
        baseUrl = `${nodeUrlProtocol}//${address.address}:${address.port}`;
    } else {
        baseUrl = `${global.location.protocol}//${global.location.hostname}`;
    }

    if (lang === 'en-US') {
        return Promise.resolve({ language: lang });
    }

    const localeJSON = await fetch(
        `${baseUrl}/static/_locales/${lang}/${langJSONFileName}`
    )
        .then(checkStatus)
        .then(parseJSON)
        .catch(err => {
            console.log(err);
            return {};
        });

    localeJSON.language = lang;
    return localeJSON;
}

export { getI18n, getLocale };
