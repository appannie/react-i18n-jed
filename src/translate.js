// @flow strict
/* eslint-disable no-redeclare */
import * as React from 'react';
import hoistStatics from 'hoist-non-react-statics';
import type { I18nType } from '.';
import { I18nContext } from './I18nProvider';

function translate<Config: { +i18n: I18nType }, Instance: ?*>(
    WrappedComponent: React.AbstractComponent<Config>
): React.AbstractComponent<$Diff<Config, { i18n: I18nType }>, Instance> & {
    WrappedComponent: React.AbstractComponent<Config>,
} {
    const name = WrappedComponent.displayName || WrappedComponent.name || 'Component';

    class Translate extends React.Component<
        $Diff<
            Config & {
                forwardedRef: *,
            },
            { i18n: I18nType }
        >
    > {
        static WrappedComponent = WrappedComponent;

        render() {
            const { forwardedRef, ...restProps } = this.props;
            return (
                <I18nContext.Consumer>
                    {i18n => (
                        <WrappedComponent ref={forwardedRef} i18n={i18n} {...restProps} />
                    )}
                </I18nContext.Consumer>
            );
        }
    }

    const ForwardedComponent = React.forwardRef<Config, Instance>((props, ref) => (
        <Translate {...props} forwardedRef={ref} />
    ));

    ForwardedComponent.displayName = `Translate(${name})`;

    return (hoistStatics(ForwardedComponent, WrappedComponent): any);
}

export default translate;
