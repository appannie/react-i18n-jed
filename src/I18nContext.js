// @flow
import * as React from 'react';
import { type I18nType } from '.';
import mockI18n from './mockI18n';

const I18nContext: React.Context<I18nType> = React.createContext(mockI18n);

export default I18nContext;
