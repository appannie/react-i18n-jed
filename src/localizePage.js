// @flow
import React from 'react';
import hoistStatics from 'hoist-non-react-statics';

import { getLocale, getI18n } from './util';
import { type I18nType, type Lang } from '.';
import I18nProvider from './I18nProvider';

const onlySingularLocaleList = ['zh-CN', 'ko-KR', 'ja-JP'];

const noop = () => {};

/* eslint gettext/no-variable-string: 0 */
export function wrapI18n(lang: Lang, i18n: I18nType): I18nType {
    const i18nAPI = {
        lang,
        gettext(text: string) {
            if (text) {
                return i18n.gettext(text);
            }
            return '';
        },
        // contextual marker
        pgettext: i18n.pgettext.bind(i18n),
        // pluralized contextually marker
        ngettext: (keyTxt: string, pluralTxt: string, count: number) => {
            // for the locale which doen't support the plural
            if (onlySingularLocaleList.includes(lang)) {
                return i18nAPI.gettext(keyTxt);
            }

            return i18n.ngettext(keyTxt, pluralTxt, count);
        },
        npgettext: (ctx: string, keyTxt: string, pluralTxt: string, count: number) => {
            // for the locale which doen't support the plural
            if (onlySingularLocaleList.includes(lang)) {
                return i18nAPI.pgettext(ctx, keyTxt);
            }

            return i18n.npgettext(ctx, keyTxt, pluralTxt, count);
        },
    };

    return i18nAPI;
}

export function localizePage(WrappedComponent: any) {
    class TranslatePage extends React.Component {
        static async getInitialProps(ctx: { req: Object }) {
            const { req } = ctx;
            const getInitialProps = WrappedComponent.getInitialProps || noop;
            const [locale: Object, wrappedComponentProps: Object] = await Promise.all([
                getLocale(req),
                getInitialProps(ctx),
            ]);

            return {
                locale,
                ...wrappedComponentProps,
            };
        }

        constructor(props, context) {
            super(props, context);

            const lang = props.locale.language;
            const i18n = getI18n(props.locale);

            this.i18n = wrapI18n(lang, i18n);
        }

        props: {
            locale: Object,
        };

        i18n: I18nType;

        render() {
            return (
                <I18nProvider i18n={this.i18n}>
                    <WrappedComponent i18n={this.i18n} {...this.props} />
                </I18nProvider>
            );
        }
    }

    TranslatePage.displayName = `TranslatePage(${WrappedComponent.displayName})`;

    return hoistStatics(TranslatePage, WrappedComponent, {
        // `getInitialProps` not in hoist static list, so we add it here to prevent override
        getInitialProps: true,
    });
}
