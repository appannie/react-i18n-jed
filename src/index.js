// @flow
import Jed, { sprintf } from 'jed';
import I18nProvider from './I18nProvider';
import translate from './translate';

export type I18nType = {
    lang: string,
    gettext: string => string,
    pgettext: (string, string) => string,
    ngettext: (string, string, number) => string,
    npgettext: (string, string, string, number) => string,
};

function getI18n(localeJSON: Object) {
    const i18n = new Jed(localeJSON);
    i18n.lang = localeJSON.language;
    return i18n;
}

export { getI18n, I18nProvider, translate, sprintf };
