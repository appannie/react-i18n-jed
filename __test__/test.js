// @flow
import React from 'react';
import { mount } from 'enzyme';
import { Jed, translate, I18nProvider, type I18nType } from '../src';
import mockI18n from '../src/mockI18n';

const localeJSON = {
    domain: 'messages',
    language: 'en-US',
    locale_data: {
        messages: {
            '': {
                domain: 'messages',
            },
            'Ad Expense': ['Test Ad Expense'],
            'App or Publisher': ['App or Publisher'],
            Cat: ['Cat', 'Cats'],
        },
    },
};

class TestElement extends React.PureComponent<{
    testProp: string,
    i18n: I18nType,
}> {
    render() {
        const { i18n } = this.props;
        return <div>{i18n.gettext('Test')}</div>;
    }
}

describe('get i18n by Jed', () => {
    it('gettext by Jed', () => {
        const i18n = new Jed(localeJSON);
        expect(i18n.gettext('Ad Expense')).toBe('Test Ad Expense');
    });
});

describe('<I18nProvider>', () => {
    it('children get i18n from I18nProvider', () => {
        const LocalizedTest = translate(TestElement);
        const WrappedTest = LocalizedTest.WrappedComponent;

        expect(<WrappedTest testProp="required" i18n={mockI18n} />).toMatchSnapshot();

        const eleWithProvider = mount(
            <I18nProvider i18n={mockI18n}>
                <LocalizedTest testProp="required" />
            </I18nProvider>
        );

        expect(eleWithProvider).toMatchSnapshot();

        const i18nFromProvider = eleWithProvider.find('TestElement').prop('i18n');
        expect(i18nFromProvider).toEqual(mockI18n);
    });
});

describe('translate Component', () => {
    it('render translated component', () => {
        const LocalizedEle = translate(TestElement);
        const localizedEle = mount(<LocalizedEle i18n={mockI18n} testProp="required" />);
        const instEle = localizedEle.instance();
        expect(instEle.props.i18n).toEqual(mockI18n);
        expect(localizedEle).toMatchSnapshot();
    });

    it('render translated stateless component', () => {
        const StatelessTest = ({
            i18n,
            testProp,
        }: {
            i18n: I18nType,
            testProp: string,
        }) => (
            <div>
                {i18n.gettext('My')}
                {testProp}
            </div>
        );

        const LocalizedEle = translate(StatelessTest);
        const localizedEle = mount(<LocalizedEle i18n={mockI18n} testProp="required" />);
        expect(localizedEle).toMatchSnapshot();
    });
});

describe('mock i18n', () => {
    it('gettext get translated message', () => {
        expect(mockI18n.gettext('txt')).toEqual('txt');
    });

    it('ngettext get translated message', () => {
        expect(mockI18n.ngettext('a', 'b', 1)).toEqual('a');
        expect(mockI18n.ngettext('a', 'b', 2)).toEqual('b');
    });

    it('pgettext get translated message', () => {
        expect(mockI18n.pgettext('ctx', 'a')).toEqual('a');
    });

    it('pgettext get translated message', () => {
        expect(mockI18n.npgettext('a', 'b', 'c', 1)).toEqual('b');
        expect(mockI18n.npgettext('a', 'b', 'c', 2)).toEqual('c');
    });
});
