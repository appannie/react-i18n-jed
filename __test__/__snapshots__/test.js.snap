// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<I18nProvider> children get i18n from I18nProvider 1`] = `
<UNDEFINED
  i18n={
    Object {
      "gettext": [Function],
      "lang": "en-US",
      "ngettext": [Function],
      "npgettext": [Function],
      "pgettext": [Function],
    }
  }
  testProp="required"
/>
`;

exports[`<I18nProvider> children get i18n from I18nProvider 2`] = `
<I18nProvider
  i18n={
    Object {
      "gettext": [Function],
      "lang": "en-US",
      "ngettext": [Function],
      "npgettext": [Function],
      "pgettext": [Function],
    }
  }
>
  <ForwardRef
    testProp="required"
  >
    <Translate(TestElement)
      forwardedRef={null}
      testProp="required"
    >
      <TestElement
        i18n={
          Object {
            "gettext": [Function],
            "lang": "en-US",
            "ngettext": [Function],
            "npgettext": [Function],
            "pgettext": [Function],
          }
        }
        testProp="required"
      >
        <div>
          Test
        </div>
      </TestElement>
    </Translate(TestElement)>
  </ForwardRef>
</I18nProvider>
`;

exports[`translate Component render anonymous component 1`] = `
<ForwardRef
  i18n={
    Object {
      "gettext": [Function],
      "lang": "en-US",
      "ngettext": [Function],
      "npgettext": [Function],
      "pgettext": [Function],
    }
  }
  testProp="required"
>
  <Translate(Component)
    forwardedRef={null}
    i18n={
      Object {
        "gettext": [Function],
        "lang": "en-US",
        "ngettext": [Function],
        "npgettext": [Function],
        "pgettext": [Function],
      }
    }
    testProp="required"
  >
    <Component
      i18n={
        Object {
          "gettext": [Function],
          "lang": "en-US",
          "ngettext": [Function],
          "npgettext": [Function],
          "pgettext": [Function],
        }
      }
      testProp="required"
    >
      <div>
        My
        required
      </div>
    </Component>
  </Translate(Component)>
</ForwardRef>
`;

exports[`translate Component render translated component 1`] = `
<Translate(TestElement)
  i18n={
    Object {
      "gettext": [Function],
      "lang": "en-US",
      "ngettext": [Function],
      "npgettext": [Function],
      "pgettext": [Function],
    }
  }
  testProp="required"
>
  <TestElement
    i18n={
      Object {
        "gettext": [Function],
        "lang": "en-US",
        "ngettext": [Function],
        "npgettext": [Function],
        "pgettext": [Function],
      }
    }
    testProp="required"
  >
    <div>
      Test
    </div>
  </TestElement>
</Translate(TestElement)>
`;

exports[`translate Component render translated stateless component 1`] = `
<ForwardRef
  i18n={
    Object {
      "gettext": [Function],
      "lang": "en-US",
      "ngettext": [Function],
      "npgettext": [Function],
      "pgettext": [Function],
    }
  }
  testProp="required"
>
  <Translate(StatelessTest)
    forwardedRef={null}
    i18n={
      Object {
        "gettext": [Function],
        "lang": "en-US",
        "ngettext": [Function],
        "npgettext": [Function],
        "pgettext": [Function],
      }
    }
    testProp="required"
  >
    <StatelessTest
      i18n={
        Object {
          "gettext": [Function],
          "lang": "en-US",
          "ngettext": [Function],
          "npgettext": [Function],
          "pgettext": [Function],
        }
      }
      testProp="required"
    >
      <div>
        My
        required
      </div>
    </StatelessTest>
  </Translate(StatelessTest)>
</ForwardRef>
`;
